### PR TITLE
dsp/fourier/internal/fftpack: use explicit complex arithmetic

### PR DESCRIPTION
**Pull request by kortschak**
It's not often that optimisations improve the legibility of the code, but these changes do.

I've made the changes in two steps, the first groups each of the complex operations together and the second then replaces the real/imag float64 operations into a single complex operation. I would recommend reviewing both of these commits rather than the change as a whole. There's a lot of change here, so please be on the look out for minor indexing errors.

Please take a look.

### CODE REVIEW

**Review by vladimir-ch — COMMENTED** [review]
In the first commit I might have skipped a line or two here and there but the third commit I think I reviewed completely. Still I hope the test cover these changes well. LGTM with two minor suggestions.

File: dsp/fourier/internal/fftpack/cfft.go [hunk 1]
```diff
@@ -404,66 +375,32 @@ func pass4(ido, l1 int, cc, ch, wa1, wa2, wa3 []float64, sign float64) {
...
-
-			ch3.set(0, k, 2, tr2-tr3)
-			ch3.set(1, k, 2, ti2-ti3)
-
-			ch3.set(0, k, 3, tr1-tr4)
-			ch3.set(1, k, 3, ti1-ti4)
+			t1 := cc3.atCmplx(0, 0, k) - cc3.atCmplx(0, 2, k)
+			t2 := cc3.atCmplx(0, 0, k) + cc3.atCmplx(0, 2, k)
+			t3 := cc3.atCmplx(0, 1, k) + cc3.atCmplx(0, 3, k)
+			t4 := scale(sign, cmplx.Conj(swap(cc3.atCmplx(0, 3, k))-swap(cc3.atCmplx(0, 1, k))))
```
[the remark this answers is not in the document]
Comment (kortschak) [reply dsp/fourier/internal/fftpack/cfft.go]: Thanks, I had intended to but missed these.

File: dsp/fourier/internal/fftpack/rfft.go [hunk 2]
```diff
@@ -304,29 +305,15 @@ func radf3(ido, l1 int, cc, ch, wa1, wa2 []float64) {
...
-			tr3 := taui * (di2 - di3)
-			ti3 := taui * (dr3 - dr2)
-
-			ch3.set(i-1, 2, k, tr2+tr3)
-			ch3.set(i, 2, k, ti2+ti3)
-
-			ch3.set(ic-1, 1, k, tr2-tr3)
-			ch3.set(ic, 1, k, ti3-ti2)
+			t2 := cc3.atCmplx(i-1, k, 0) + scale(taur, c2)
+			t3 := scale(taui, cmplx.Conj(swap(d2)-swap(d3)))
```
[the remark this answers is not in the document]
Comment (kortschak) [reply dsp/fourier/internal/fftpack/rfft.go]: Done

**Comment by kortschak** [comment]
Thanks. It is well covered, but not explicitly and obviously so; the tests in dsp/fourier test this extensively, but the tests in fftpack itself don't. I will fix this in this PR, but I'll take a break first (this hurt due to the tedium). I propose to add a naive fft and ifft and reflect the same range of tests that dsp/fourier does (it's necessary to iterate over a fairly large range of sequence/coefficient length to hit all the code paths).

This is necessary for the next sweeps of optimisations which will remove the last traces of Fortran indexing and then make the work slices be `[]complex128`. This will reduce further the runtime bounds checking costs, but it's scary, so having better tests will be important.

**Comment by kortschak** [comment]
Oh, happy. Coverage is better than I though cfft is 92.9% and rfft is 90.5%. Still will follow the plan proposed though.

**Comment by kortschak** [comment]
Coverage increased to all functions for cfft (96.9%) and rfft (98.2%).

PTAL

File: dsp/fourier/internal/fftpack/cfft.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (kortschak) [reply dsp/fourier/internal/fftpack/cfft.go]: Looking into this more carefully after having had a break, there is merit to using both forms; conj(swap(x)) == -xi while swap(conj(x)) == xi. The choice impacts on the apparent structure of the transform and in this case (and L395), I think that the structure is more visible. However, it does look at hoc. What I think might be worth doing is adding wrapper functions that state what is happening,
```
func i(c complex128) complex128 {
	return swap(cmplx.Conj(c))
}

func negi(c complex128) complex128 {
	return cmplx.Conj(swap(c))
}
```
This should make the code easier to read and the intention obvious. What do you think?

File: dsp/fourier/internal/fftpack/fftpack_test.go [hunk 3]
```diff
@@ -14,116 +14,169 @@ import (
...
+	}
+	if wantiifac != nil && !reflect.DeepEqual(ifac, wantiifac) {
+		failed = true
+		t.Errorf("unexpected ifac after call to rffti for n=%d", n)
+	}
+	if failed {
+		return
+	}
+
+	// Construct a sequence with known a frequency spectrum and compare
```
Comment (vladimir-ch) [inline dsp/fourier/internal/fftpack/fftpack_test.go:56]: nit: a known frequency

File: dsp/fourier/internal/fftpack/fftpack_test.go [hunk 4]
```diff
@@ -14,116 +14,169 @@ import (
...
+	// the computed spectrum.
+	modn := n % 2
+	fn := float64(n)
+	nm1 := n - 1
```
Comment (vladimir-ch) [inline dsp/fourier/internal/fftpack/fftpack_test.go:60]: This variable is used at only one place but it seems that this comes from the original Fortran code and I guess you plan to address these Fortranisms all at once.

**Review by vladimir-ch — APPROVED** [review]

File: dsp/fourier/internal/fftpack/fftpack_test.go (hunk 4 above)
Comment (kortschak) [reply dsp/fourier/internal/fftpack/fftpack_test.go:60]: Same here.

File: dsp/fourier/internal/fftpack/cfft.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (vladimir-ch) [reply dsp/fourier/internal/fftpack/cfft.go]: SGTM

**Pull request merged**